technology-considerations/graphql-prototypes.md

```diff
@@ -0,0 +1,193 @@
+and the Lambda code together, or give up on local development and only run code
+in an AWS environment.
+
+One final reason disadvantage of AppSync is that it requires you to write Apache
+Velocity templates. One of the developers on the team has used Velocity in
+a previous project and found it very difficult to read, maintain and test.
+
+[tdr-appsync-prototype]: https://github.com/nationalarchives/prototype-appsync
+[serverless]: https://serverless.com/
+[serverless-java]: https://serverless.com/blog/how-to-create-a-rest-api-in-java-using-dynamodb-and-serverless/
+[amplify]: https://github.com/aws-amplify/amplify-cli
+
+### Deployment
+
+The Serverless Framework and Amplify both depend on [CloudFormation], which
+would be tricky to fit in with our [Terraform configuration][tdr-terraform].
+
+If we built our own deployment pipeline, it would have to deploy the Velocity
+templates using the AWS API as well as deploying the Lambda code by copying
+build artifacts.
+
+[CloudFormation]: https://aws.amazon.com/cloudformation/
+[tdr-terraform]: https://github.com/nationalarchives/tdr-prototype-terraform
+
+### Hosting
+
+Using AppSync would make the application extremely coupled to AWS
+Infrastructure, because so much of the code would be in Velocity templates and
+the rest would be implemented as AWS Lambdas.
+
+### Database connection
+
+It's straightforward to connect AppSync to DynamoDB, but we'd [prefer to use a
+relational DB][db-choice] to store information about transfers.
+
+To connect AppSync to a relational DB such as Aurora, you can either use a [Data
+API][rds-data-api] or use a Lambda as a data source, both of which have
+downsides.
+
+[Data API is not currently available in eu-west-2][data-api-availability], and
+we need to store data in the UK.
+
+Lambda data sources suffer from cold starts, which may be even slower in a VPC
+(although [AWS say this will improve][lambda-vpc-improvements]).
+
+[rds-data-api]: https://docs.aws.amazon.com/appsync/latest/devguide/tutorial-rds-resolvers.html
+[data-api-availability]: https://docs.aws.amazon.com/AmazonRDS/latest/AuroraUserGuide/data-api.html#data-api.regions
+[lambda-vpc-improvements]: https://www.nuweba.com/AWS-Lambda-in-a-VPC-will-soon-be-faster
+
+## Sangria approach
+
+As mentioned above, we prefer to use Scala as a backend language for consistency
+with our other projects. [Sangria] looks like a mature and feature-rich
+implementation of a GraphQL server in Scala, so we are using it to build a
+[GraphQL API prototype][sangria-prototype]. It uses [Slick] to map the Scala
+code to SQL queries.
+
+### Local development
+
+The development process is simple because it's just a single Scala application.
+We've wrapped the Sangria code in an [akka-http] server for development, which
+is quick to redeploy when modifying the code.
+
+We've added several API endpoints so far, and it's been fairly straightforward
+to build everything we need.
+
+### Deployment and hosting
+
```

Comment:
We're not running this in a lambda any more so the akka-http server is what we're using in the dev environment on ECS as well as for local development and we're not using the lambda package at all.

Reply:
I've updated this section.